addons/mass_mailing/controllers/main.py

```diff
@@ -62,6 +62,11 @@ def mailing(self, mailing_id, email=None, res_id=None, token="", **post):
                     'show_blacklist_button': request.env['ir.config_parameter'].sudo().get_param('mass_mailing.show_blacklist_buttons'),
                 })
             else:
+                contacts = request.env['mail.mass_mailing.contact'].sudo().search([('email', '=', email)])
```

Comment:
Hello,

This code could be simplified, notably I think by lessening number of search / browse. Indeed you can do a search directly on the subscription model and use the result to make a record set. It could be something like

```
opt_in_lists = self.env['mail.mass_mailing.list_contact_rel'].sudo().search([
    ('contact_id.email', '=', email),
    ('opt_out', '=', False)
]).mapped('list_id')
```
This makes the code easier to read and understand.

By the way as mapped and filter return record sets you don't have to get the IDs and re-do a browse. Moreover the environment is propagated, meaning that doing a search in sudo and working on the result gives you a result still in the superuser environment.

Does that reads feasible ?